### ISSUE TITLE
How can I do batch upserts with Ktorm?

### ISSUE THREAD

**Issue by PedroD** [kotlin-orm/ktorm#220]
Hello!

First of all, thanks for providing the world with such a great project! 🥳

I have one question:

How can I implement this same Postgres query using Ktorm DSL?

```SQL
INSERT INTO books (title, author, last_edition_number, number_of_prints)
VALUES 
(:title, :author, :last_edition_number, :number_of_prints)
ON CONFLICT (title, author) -- We don't allow duplicated books by author
DO UPDATE SET
        last_edition_number = EXCLUDED.last_edition_number, 
        number_of_prints= EXCLUDED.number_of_prints
;
```

> Basically what this query does is to insert a new book, but if a book with the same combination of title and author exists, instead of returning an error, updates the existing record's fields last_edition_number and number_of_prints.

And then, after implementing it in the DSL, how can I run this in batch for a list of books I want to upsert?

**Comment by vincentlauvlwj** [comment]
If you are using MySQL, you can use the `bulkInsert` function: 

```kotlin
database.bulkInsert(Employees) {
    item {
        set(it.name, "jerry")
        set(it.job, "trainee")
        set(it.managerId, 1)
        set(it.hireDate, LocalDate.now())
        set(it.salary, 50)
        set(it.departmentId, 1)
    }
    item {
        set(it.name, "linda")
        set(it.job, "assistant")
        set(it.managerId, 3)
        set(it.hireDate, LocalDate.now())
        set(it.salary, 100)
        set(it.departmentId, 2)
    }
    onDuplicateKey {
        set(it.name, it.name)
        set(it.job, values(it.job))
    }
}
```

Generated SQL: 

```mysql
INSERT INTO t_employee (name, job, manager_id, hire_date, salary, department_id) 
VALUES (?, ?, ?, ?, ?, ?), (?, ?, ?, ?, ?, ?) 
ON DUPLICATE KEY UPDATE name = name, job = VALUES(job) 
```

But, unfortunately, we don't have such function for PostgreSQL yet. You can read the source code [here](https://github.com/kotlin-orm/ktorm/blob/master/ktorm-support-mysql/src/main/kotlin/org/ktorm/support/mysql/BulkInsert.kt#L85-L93) and try to implement the PostgreSQL's version by your self. 

BTW, after doing that, please send me a PR if you like, looking forward to your contribution.

**Comment by PedroD** [comment]
Cool! I'll have a look and see what I can do!

**Comment by PedroD** [comment]
@vincentlauvlwj I don't understand this test in the PostgreSQL:

```kotlin
    @Test
    fun testInsertOrUpdate() {
        database.insertOrUpdate(Employees) {
            set(it.id, 1)
            set(it.name, "vince")
            set(it.job, "engineer")
            set(it.salary, 1000)
            set(it.hireDate, LocalDate.now())
            set(it.departmentId, 1)

            onDuplicateKey {
                set(it.salary, it.salary + 900)
            }
        }
        database.insertOrUpdate(Employees) {
            set(it.id, 5)
            set(it.name, "vince")
            set(it.job, "engineer")
            set(it.salary, 1000)
            set(it.hireDate, LocalDate.now())
            set(it.departmentId, 1)

            onDuplicateKey(it.id) {
                set(it.salary, it.salary + 900)
            }
        }

        assert(database.employees.find { it.id eq 1 }!!.salary == 1000L)
        assert(database.employees.find { it.id eq 5 }!!.salary == 1000L)
    }
```

Shouldn't the asserts be:

assert(database.employees.find { it.id eq 1 }!!.salary == 1**900**L)
assert(database.employees.find { it.id eq 5 }!!.salary == 1**900**L)

?

**Comment by PedroD** [comment]
@vincentlauvlwj https://github.com/kotlin-orm/ktorm/pull/226

**Comment by vincentlauvlwj** [comment]
Implemented in Ktorm 3.3.0.
